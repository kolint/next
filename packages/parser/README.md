---
name: parser
---

# Parser

<!-- @include docs/parts/packages/parser/description.md-->

Parses HTML views with [Knockout] bindings.

<!-- /include -->

<!-- @include docs/parts/package-nav.md -->

[**Documentation**](https://kolint.github.io/next) | [Package (npm)](https://npmjs.com/package/@kolint/parser) | [Repository](https://github.com/kolint/next) | [Source Code](https://github.com/kolint/next/tree/main/packages/parser)

<!-- /include -->

<!-- @include docs/parts/reference.md -->

[TypeScript]: https://typescriptlang.org
[ESLint]: https://eslint.org
[Knockout]: https://knockoutjs.com
[toolchain]: https://kolint.github.io/next

<!-- /include -->
