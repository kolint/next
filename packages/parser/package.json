{
  "name": "@kolint/parser",
  "version": "0.1.0-beta.1",
  "type": "module",
  "exports": "./src/index.ts",
  "dependencies": {
    "parse5": "^7.1.2",
    "acorn": "^8.11.3"
  },
  "publishConfig": {
    "access": "public"
  }
}
