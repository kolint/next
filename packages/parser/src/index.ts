export * from "./parse5-utils.js";
export * from "./parser.js";
