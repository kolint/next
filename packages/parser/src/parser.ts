import {
  isParse5CommentNode,
  isParse5Element,
  isParse5TextNode,
  parse5,
  parse5LocationToRange,
  type parse5TreeAdapter,
} from "./parse5-utils.js";
import { Range, Position } from "@kolint/location";
import {
  Document,
  type Node,
  Text,
  VirtualElement,
  Comment,
  Element,
  Attribute,
  Binding,
  Scope,
} from "@kolint/syntax-tree";
import * as acorn from "acorn";

const virtualElementStart = /^(\s*(#?)ko\s+)([^\s]+)(\s*:\s*)([^]*?)(\s*)$/;
const virtualElementEnd = /^\s*\/ko\s[^]*$/;

export interface ParseOptions {
  bindingAttributes?: readonly string[];
  onError?: ((error: parse5.ParserError) => void) | undefined;
}

export function parse(document: string, options?: ParseOptions) {
  const parser = new Parser(document, options);
  return parser.parse();
}

class Parser {
  #bindingAttributes: readonly string[];
  #onError: ((error: parse5.ParserError) => void) | undefined;
  #source: string;

  constructor(source: string, options?: ParseOptions) {
    this.#source = source;
    this.#bindingAttributes = options?.bindingAttributes ?? ["data-bind"];
    this.#onError = options?.onError;
  }

  parse(): Document {
    const root = parse5.parseFragment(this.#source, {
      sourceCodeLocationInfo: true,
      scriptingEnabled: false,
      onParseError: this.#onError,
    });
    const iter = root.childNodes[Symbol.iterator]();
    const children: Node[] = [];
    let result: IteratorResult<parse5TreeAdapter.Node> | undefined;

    while (!(result = iter.next()).done) {
      children.push(this.#parseNode(result.value, iter));
    }

    return new Document(children, new Range(Position.zero, Position.zero));
  }

  #parseNode(
    node: parse5TreeAdapter.Node,
    iter: Iterator<parse5TreeAdapter.Node>,
  ): Node {
    switch (true) {
      case isParse5TextNode(node): {
        return new Text(
          node.value,
          parse5LocationToRange(node.sourceCodeLocation!),
        );
      }
      case isParse5CommentNode(node): {
        if (virtualElementStart.test(node.data)) {
          const [
            spacing1,
            prefix,
            bindingName,
            _spacing2,
            bindingParam,
            spacing3,
          ] = virtualElementStart.exec(node.data)!.slice(1) as [
            string,
            string,
            string,
            string,
            string,
            string,
          ];

          const hidden = prefix === "#";

          let balance = 1;
          const children: parse5TreeAdapter.Node[] = [];
          let result: IteratorResult<parse5TreeAdapter.Node> | undefined;
          let endComment: parse5TreeAdapter.CommentNode | undefined;

          while (!(result = iter.next()).done) {
            if (isParse5CommentNode(result.value)) {
              if (virtualElementStart.test(result.value.data)) {
                ++balance;
              } else if (virtualElementEnd.test(result.value.data)) {
                --balance;

                if (balance === 0) {
                  endComment = result.value as parse5TreeAdapter.CommentNode;
                  break;
                }
              }
            }

            children.push(result.value);
          }

          if (!endComment) {
            throw new Error("Unbalanced virtual element (knockout comment).");
          }

          const iter2 = children[Symbol.iterator]();
          const children2: Node[] = [];
          let result2: IteratorResult<parse5TreeAdapter.Node> | undefined;

          while (!(result2 = iter2.next()).done) {
            children2.push(this.#parseNode(result2.value, iter2));
          }

          const virtualElement = new VirtualElement(
            undefined!,
            hidden,
            children2,
            new Comment(
              node.data,
              parse5LocationToRange(node.sourceCodeLocation!),
            ),
            new Comment(
              endComment.data,
              parse5LocationToRange(endComment.sourceCodeLocation!),
            ),
            new Range(
              parse5LocationToRange(node.sourceCodeLocation!).start,
              parse5LocationToRange(endComment.sourceCodeLocation!).end,
            ),
          );

          // Hack self reference
          (virtualElement as { binding: Binding }).binding = new Binding(
            new Scope(
              bindingName,
              new Range(
                Position.fromOffset(
                  node.sourceCodeLocation!.startOffset +
                    "<!--".length +
                    spacing1.length,
                  this.#source,
                ),
                Position.fromOffset(
                  node.sourceCodeLocation!.startOffset +
                    "<!--".length +
                    spacing1.length +
                    bindingName.length,
                  this.#source,
                ),
              ),
            ),
            new Scope(
              bindingParam,
              new Range(
                Position.fromOffset(
                  node.sourceCodeLocation!.endOffset -
                    bindingParam.length -
                    spacing3.length -
                    "-->".length,
                  this.#source,
                ),
                Position.fromOffset(
                  node.sourceCodeLocation!.endOffset -
                    spacing3.length -
                    "-->".length,
                  this.#source,
                ),
              ),
            ),
            virtualElement,
            parse5LocationToRange(node.sourceCodeLocation!),
          );

          return virtualElement;
        } else {
          return new Comment(
            node.data,
            parse5LocationToRange(node.sourceCodeLocation!),
          );
        }
      }
      case isParse5Element(node): {
        const iter = node.childNodes[Symbol.iterator]();
        const children: Node[] = [];
        let current: IteratorResult<parse5TreeAdapter.Node> | undefined;

        while (!(current = iter.next()).done) {
          children.push(this.#parseNode(current.value, iter));
        }

        const attributes = node.attrs.map((attr) => {
          const _quote = this.#source.at(
            node.sourceCodeLocation!.attrs![attr.name]!.startOffset +
              attr.name.length +
              1,
          );
          const quoted = (["'", '"'] as unknown[]).includes(_quote);
          const quote = quoted ? (_quote as '"' | "'") : null;

          const range = parse5LocationToRange(
            node.sourceCodeLocation!.attrs![attr.name]!,
          );

          return new Attribute(
            new Scope(
              attr.name,
              new Range(
                range.start,
                Position.fromOffset(
                  range.start.offset + attr.name.length,
                  this.#source,
                ),
              ),
            ),
            new Scope(
              attr.value,
              new Range(
                Position.fromOffset(
                  range.start.offset +
                    // name
                    attr.name.length +
                    // equal
                    1 +
                    // quote
                    (quoted ? 1 : 0),
                  this.#source,
                ),
                range.end,
              ),
            ),
            attr.namespace,
            attr.prefix,
            quote,
            range,
          );
        });

        const bindings = attributes
          .filter((attr) => this.#bindingAttributes.includes(attr.name.text))
          .flatMap((attr): Binding[] => {
            const expressionText = `({${attr.value.text}})`;
            const expression = acorn.parseExpressionAt(expressionText, 0, {
              ecmaVersion: "latest",
              sourceType: "script",
              ranges: true,
            });
            if (expression.type !== "ObjectExpression") {
              throw new Error("Expected ObjectExpression.");
            }

            const translate = (offset: number) =>
              // name
              attr.name.range.end.offset +
              // quote
              (attr.quote ? 1 : 0) +
              // offset
              offset -
              // TODO: No idea why?
              1;

            return expression.properties.map((prop) => {
              if (prop.type === "SpreadElement") {
                throw new Error("Spread syntax is not supported in bindings.");
              }

              if (prop.computed) {
                throw new Error(
                  "Computed property as binding is not supported.",
                );
              }

              let name: string;

              if (prop.key.type === "Identifier") {
                name = prop.key.name;
              } else if (prop.key.type === "Literal" && prop.key.raw) {
                name = prop.key.raw;
              } else {
                throw new Error("Unsupported property key in binding.");
              }

              const param = expressionText.slice(
                translate(prop.value.range![0]),
                translate(prop.value.range![1]),
              );

              return new Binding(
                new Scope(
                  name,
                  new Range(
                    Position.fromOffset(
                      translate(prop.key.range![0]),
                      this.#source,
                    ),
                    Position.fromOffset(
                      translate(prop.key.range![1]),
                      this.#source,
                    ),
                  ),
                ),
                new Scope(
                  param,
                  new Range(
                    Position.fromOffset(
                      translate(prop.value.range![0]),
                      this.#source,
                    ),
                    Position.fromOffset(
                      translate(prop.value.range![1]),
                      this.#source,
                    ),
                  ),
                ),
                attr,
                new Range(
                  Position.fromOffset(translate(prop.range![0]), this.#source),
                  Position.fromOffset(translate(prop.range![1]), this.#source),
                ),
              );
            });
          });

        return new Element(
          node.tagName,
          attributes,
          bindings,
          children,
          parse5LocationToRange(node.sourceCodeLocation!),
        );
      }
      default: {
        throw new Error("Unexpected node type");
      }
    }
  }
}
