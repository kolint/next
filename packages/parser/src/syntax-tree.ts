import type { Range } from "./location.js";

export abstract class Node {
  constructor(public range: Range) {}
}

export class Text extends Node {
  constructor(
    public readonly content: string,
    range: Range,
  ) {
    super(range);
  }
}

export class Comment extends Node {
  constructor(
    public readonly content: string,
    range: Range,
  ) {
    super(range);
  }
}

export class Scope {
  constructor(
    public readonly text: string,
    public readonly range: Range,
  ) {}
}

export class Attribute {
  constructor(
    public readonly name: Scope,
    public readonly value: Scope,
    public readonly namespace: string | undefined,
    public readonly prefix: string | undefined,
    public readonly quote: "'" | '"' | null,
    public readonly range: Range,
  ) {}
}

export class Binding {
  constructor(
    public readonly name: Scope,
    public readonly param: Scope,
    public readonly parent: Attribute | VirtualElement,
    public readonly range: Range,
  ) {}
}

export class Element extends Node {
  constructor(
    public readonly tagName: string,
    public readonly attributes: readonly Attribute[],
    public readonly bindings: readonly Binding[],
    public readonly children: readonly Node[],
    range: Range,
  ) {
    super(range);
  }
}

export class VirtualElement extends Node {
  constructor(
    public readonly binding: Binding,
    public readonly hidden: boolean,
    public readonly children: readonly Node[],
    public readonly start: Comment,
    public readonly end: Comment,
    range: Range,
  ) {
    super(range);
  }
}

export class Document extends Node {
  constructor(
    public readonly children: readonly Node[],
    range: Range,
  ) {
    super(range);
  }
}

export type ChildNode = Element | VirtualElement | Text | Comment;

export function isChildNode(node: Node): node is ChildNode {
  return (
    node instanceof Element ||
    node instanceof VirtualElement ||
    node instanceof Text ||
    node instanceof Comment
  );
}

export type ParentNode = Element | VirtualElement | Document;

export function isParentNode(node: Node): node is ParentNode {
  return (
    node instanceof Element ||
    node instanceof VirtualElement ||
    node instanceof Document
  );
}
