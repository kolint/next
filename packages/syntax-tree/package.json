{
  "name": "@kolint/syntax-tree",
  "version": "0.1.0",
  "type": "module",
  "exports": "./src/index.ts",
  "dependencies": {
    "@kolint/location": "workspace:^"
  },
  "engines": {
    "node": ">=20.0.0"
  }
}
