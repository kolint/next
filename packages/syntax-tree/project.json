{
  "name": "@kolint/syntax-tree",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "projectType": "library",
  "sourceRoot": "{projectRoot}/src",
  "targets": {
    "build": {},
    "test": {},
    "lint": {}
  }
}
