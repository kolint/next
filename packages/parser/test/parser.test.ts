import { parse } from "../src/parser.js";
import { VirtualElement } from "@kolint/syntax-tree";
import { describe, test, expect } from "bun:test";
import assert from "node:assert/strict";

describe("parser", () => {
  test("Deep virtual elements", () => {
    const document = parse(
      "<!-- ko foo: foo --><!-- ko bar: bar --><!-- /ko --><!-- /ko -->",
    );
    expect(document).toMatchSnapshot();
  });

  test("Hidden virtual element", () => {
    const document = parse("<!-- #ko foo: foo --><!-- /ko -->");
    expect(document).toMatchSnapshot();
    assert(document.children[0] instanceof VirtualElement);
    expect(document.children[0].hidden).toBe(true);
  });

  test("Element bindings", () => {
    const document = parse("<div data-bind='0: bar'></div>");
    expect(document).toMatchSnapshot();
  });
});
