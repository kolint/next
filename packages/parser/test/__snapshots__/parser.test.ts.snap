// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parser Deep virtual elements 1`] = `
Document {
  "children": [
    VirtualElement {
      "binding": Binding {
        "name": Scope {
          "range": Range {
            "end": Position {
              "column": 11,
              "line": 0,
              "offset": 11,
            },
            "start": Position {
              "column": 8,
              "line": 0,
              "offset": 8,
            },
          },
          "text": "foo",
        },
        "param": Scope {
          "range": Range {
            "end": Position {
              "column": 16,
              "line": 0,
              "offset": 16,
            },
            "start": Position {
              "column": 13,
              "line": 0,
              "offset": 13,
            },
          },
          "text": "foo",
        },
        "parent": [Circular],
        "range": Range {
          "end": Position {
            "column": 20,
            "line": 0,
            "offset": 20,
          },
          "start": Position {
            "column": 0,
            "line": 0,
            "offset": 0,
          },
        },
      },
      "children": [
        VirtualElement {
          "binding": Binding {
            "name": Scope {
              "range": Range {
                "end": Position {
                  "column": 31,
                  "line": 0,
                  "offset": 31,
                },
                "start": Position {
                  "column": 28,
                  "line": 0,
                  "offset": 28,
                },
              },
              "text": "bar",
            },
            "param": Scope {
              "range": Range {
                "end": Position {
                  "column": 36,
                  "line": 0,
                  "offset": 36,
                },
                "start": Position {
                  "column": 33,
                  "line": 0,
                  "offset": 33,
                },
              },
              "text": "bar",
            },
            "parent": [Circular],
            "range": Range {
              "end": Position {
                "column": 40,
                "line": 0,
                "offset": 40,
              },
              "start": Position {
                "column": 20,
                "line": 0,
                "offset": 20,
              },
            },
          },
          "children": [],
          "end": Comment {
            "content": " /ko ",
            "range": Range {
              "end": Position {
                "column": 52,
                "line": 0,
                "offset": 52,
              },
              "start": Position {
                "column": 40,
                "line": 0,
                "offset": 40,
              },
            },
          },
          "hidden": false,
          "range": Range {
            "end": Position {
              "column": 52,
              "line": 0,
              "offset": 52,
            },
            "start": Position {
              "column": 20,
              "line": 0,
              "offset": 20,
            },
          },
          "start": Comment {
            "content": " ko bar: bar ",
            "range": Range {
              "end": Position {
                "column": 40,
                "line": 0,
                "offset": 40,
              },
              "start": Position {
                "column": 20,
                "line": 0,
                "offset": 20,
              },
            },
          },
        },
      ],
      "end": Comment {
        "content": " /ko ",
        "range": Range {
          "end": Position {
            "column": 64,
            "line": 0,
            "offset": 64,
          },
          "start": Position {
            "column": 52,
            "line": 0,
            "offset": 52,
          },
        },
      },
      "hidden": false,
      "range": Range {
        "end": Position {
          "column": 64,
          "line": 0,
          "offset": 64,
        },
        "start": Position {
          "column": 0,
          "line": 0,
          "offset": 0,
        },
      },
      "start": Comment {
        "content": " ko foo: foo ",
        "range": Range {
          "end": Position {
            "column": 20,
            "line": 0,
            "offset": 20,
          },
          "start": Position {
            "column": 0,
            "line": 0,
            "offset": 0,
          },
        },
      },
    },
  ],
  "range": Range {
    "end": Position {
      "column": 0,
      "line": 0,
      "offset": 0,
    },
    "start": Position {
      "column": 0,
      "line": 0,
      "offset": 0,
    },
  },
}
`;

exports[`parser Hidden virtual element 1`] = `
Document {
  "children": [
    VirtualElement {
      "binding": Binding {
        "name": Scope {
          "range": Range {
            "end": Position {
              "column": 12,
              "line": 0,
              "offset": 12,
            },
            "start": Position {
              "column": 9,
              "line": 0,
              "offset": 9,
            },
          },
          "text": "foo",
        },
        "param": Scope {
          "range": Range {
            "end": Position {
              "column": 17,
              "line": 0,
              "offset": 17,
            },
            "start": Position {
              "column": 14,
              "line": 0,
              "offset": 14,
            },
          },
          "text": "foo",
        },
        "parent": [Circular],
        "range": Range {
          "end": Position {
            "column": 21,
            "line": 0,
            "offset": 21,
          },
          "start": Position {
            "column": 0,
            "line": 0,
            "offset": 0,
          },
        },
      },
      "children": [],
      "end": Comment {
        "content": " /ko ",
        "range": Range {
          "end": Position {
            "column": 33,
            "line": 0,
            "offset": 33,
          },
          "start": Position {
            "column": 21,
            "line": 0,
            "offset": 21,
          },
        },
      },
      "hidden": true,
      "range": Range {
        "end": Position {
          "column": 33,
          "line": 0,
          "offset": 33,
        },
        "start": Position {
          "column": 0,
          "line": 0,
          "offset": 0,
        },
      },
      "start": Comment {
        "content": " #ko foo: foo ",
        "range": Range {
          "end": Position {
            "column": 21,
            "line": 0,
            "offset": 21,
          },
          "start": Position {
            "column": 0,
            "line": 0,
            "offset": 0,
          },
        },
      },
    },
  ],
  "range": Range {
    "end": Position {
      "column": 0,
      "line": 0,
      "offset": 0,
    },
    "start": Position {
      "column": 0,
      "line": 0,
      "offset": 0,
    },
  },
}
`;

exports[`parser Element bindings 1`] = `
Document {
  "children": [
    Element {
      "attributes": [
        Attribute {
          "name": Scope {
            "range": Range {
              "end": Position {
                "column": 14,
                "line": 0,
                "offset": 14,
              },
              "start": Position {
                "column": 5,
                "line": 0,
                "offset": 5,
              },
            },
            "text": "data-bind",
          },
          "namespace": undefined,
          "prefix": undefined,
          "quote": "'",
          "range": Range {
            "end": Position {
              "column": 23,
              "line": 0,
              "offset": 23,
            },
            "start": Position {
              "column": 5,
              "line": 0,
              "offset": 5,
            },
          },
          "value": Scope {
            "range": Range {
              "end": Position {
                "column": 23,
                "line": 0,
                "offset": 23,
              },
              "start": Position {
                "column": 16,
                "line": 0,
                "offset": 16,
              },
            },
            "text": "0: bar",
          },
        },
      ],
      "bindings": [
        Binding {
          "name": Scope {
            "range": Range {
              "end": Position {
                "column": 17,
                "line": 0,
                "offset": 17,
              },
              "start": Position {
                "column": 16,
                "line": 0,
                "offset": 16,
              },
            },
            "text": "0",
          },
          "param": Scope {
            "range": Range {
              "end": Position {
                "column": 22,
                "line": 0,
                "offset": 22,
              },
              "start": Position {
                "column": 19,
                "line": 0,
                "offset": 19,
              },
            },
            "text": "bar",
          },
          "parent": Attribute {
            "name": Scope {
              "range": Range {
                "end": Position {
                  "column": 14,
                  "line": 0,
                  "offset": 14,
                },
                "start": Position {
                  "column": 5,
                  "line": 0,
                  "offset": 5,
                },
              },
              "text": "data-bind",
            },
            "namespace": undefined,
            "prefix": undefined,
            "quote": "'",
            "range": Range {
              "end": Position {
                "column": 23,
                "line": 0,
                "offset": 23,
              },
              "start": Position {
                "column": 5,
                "line": 0,
                "offset": 5,
              },
            },
            "value": Scope {
              "range": Range {
                "end": Position {
                  "column": 23,
                  "line": 0,
                  "offset": 23,
                },
                "start": Position {
                  "column": 16,
                  "line": 0,
                  "offset": 16,
                },
              },
              "text": "0: bar",
            },
          },
          "range": Range {
            "end": Position {
              "column": 22,
              "line": 0,
              "offset": 22,
            },
            "start": Position {
              "column": 16,
              "line": 0,
              "offset": 16,
            },
          },
        },
      ],
      "children": [],
      "range": Range {
        "end": Position {
          "column": 30,
          "line": 0,
          "offset": 30,
        },
        "start": Position {
          "column": 0,
          "line": 0,
          "offset": 0,
        },
      },
      "tagName": "div",
    },
  ],
  "range": Range {
    "end": Position {
      "column": 0,
      "line": 0,
      "offset": 0,
    },
    "start": Position {
      "column": 0,
      "line": 0,
      "offset": 0,
    },
  },
}
`;
